rfcs/rfc-bld-20220623-1-build-failure-rca.md

```diff
@@ -0,0 +1,167 @@
+    "type" : "STAGE",
+    "edges" : [
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "1718",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+The RCA dispatcher lambda function uses these pipeline stage data to build a relationship graph of build platform/configuration, and then it triggers the RCA worker lambda function to do build failure RCA for failed build platform/configuration.
+
+RCA worker lambda function sends a GET request
+```
+https://{JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/{JOB_NAME}/runs/{BUILD_NUMBER}/nodes/{ID}/log
+```
+to get the build log of a certain build platform/configuration.
+#### Build Failure RCA Patterns
+RCA worker lambda function scans the build log and search for log messages that match build failure RCA patterns.
+
+Current build RCA patterns from https://github.com/aws-lumberyard/build-failure-rca/tree/main/build-failure-rca/rca-patterns can be reused.
+
+#### Workflow
+![workflow](rfc-bld-20220623-1-build-failure-rca/workflow.png)
+
+1. A change in build failure RCA repo triggers the build failure RCA CDK deployment pipeline.
+2. The build failure RCA CDK deployment pipeline deploys build failure RCA CDK stack to AWS accounts that are using build failure RCA system.
+3. Change in O3DE repo or O3DE fork repo triggers the AR or Periodic builds.
+4. After each AR or Periodic build completes, it publishes build data to post build SNS topic.
+5. Post build SNS topic triggers the RCA dispatcher lambda function.
+6. RCA dispatcher lambda function retrieves AR/Periodic build data and dispatch the actual RCA work for each build platform/configuration to separate RCA worker lambda functions. 
+7. RCA worker lambda function retrieves build log of a certain build platform/configuration, it scans the build log and search for log messages that match build failure RCA patterns, once the RCA work is done, it uploads the build log to S3 and return RCA result to RCA dispatcher lambda function.
+8. Once RCA dispatcher received the RCA result from RCA worker, it uploads the RCA result to CloudWatch metrics.
+9. Once RCA dispatcher received the RCA result from RCA worker, it sends build failure notifications via email, Discord or Slack.
+
+### What are the advantages of the feature?
+- It offloads build failure RCA work from Jenkins server to lambda functions, and thus reduce loads on Jenkins server.
+- Build failure RCA system can be deployed by CDK, it can be easily deployed to different AWS accounts that are willing to use build failure RCA system.
+- Build failure RCA is done per build platform/configuration instead of whole AR/Perodic pipeline.
+- Build logs can be kept longer on S3.
+- More detailed build failure RCA result can be included in build failure notification.
+
+### What are the disadvantages of the feature?
+- This is not directly integrated with Jenkins, and thus the build failure RCA result cannot be shown on Jenkins UI.
```

Comment:
Might be a way to show html badges in the job summary, so you can link to failures, though it's done with yet another plugin: https://plugins.jenkins.io/badge/

Reply:
This plugin looks good, I will do some investigations.